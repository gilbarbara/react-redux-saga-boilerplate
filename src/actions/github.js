// @flow
/**
 * @module Actions/User
 * @desc User Actions
 */
import { createActions } from 'redux-actions';

import { ActionTypes } from 'constants/index';

export const { githubGetRepos: getRepos } = createActions({
  [ActionTypes.GITHUB_GET_REPOS]: (query: string) => ({ query }),
});
