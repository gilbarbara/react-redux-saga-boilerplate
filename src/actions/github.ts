/**
 * @module Actions/User
 * @desc User Actions
 */
import { createAction } from 'modules/helpers';

import { ActionTypes } from 'literals';

export const getRepos = createAction(ActionTypes.GITHUB_GET_REPOS_REQUEST, (query: string) => ({
  query,
}));
