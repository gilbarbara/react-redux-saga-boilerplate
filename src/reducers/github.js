import { handleActions } from 'redux-actions';
import immutable from 'immutability-helper';
import { parseError } from 'modules/client';

import { ActionTypes, STATUS } from 'constants/index';

export const githubState = {
  repos: {
    data: {},
    status: STATUS.IDLE,
    message: '',
    query: '',
  },
};

export default {
  github: handleActions(
    {
      [ActionTypes.GITHUB_GET_REPOS]: (state, { payload }) => {
        const data = state.repos.data[payload.query] ? state.repos.data[payload.query] : [];

        return immutable(state, {
          repos: {
            data: {
              [payload.query]: { $set: data },
            },
            message: { $set: '' },
            query: { $set: payload.query },
            status: { $set: STATUS.RUNNING },
          },
        });
      },
      [ActionTypes.GITHUB_GET_REPOS_SUCCESS]: (state, { payload }) =>
        immutable(state, {
          repos: {
            data: {
              [state.repos.query]: { $set: payload.data || [] },
            },
            status: { $set: STATUS.READY },
          },
        }),
      [ActionTypes.GITHUB_GET_REPOS_FAILURE]: (state, { payload }) =>
        immutable(state, {
          repos: {
            message: { $set: parseError(payload.message) },
            status: { $set: STATUS.ERROR },
          },
        }),
    },
    githubState,
  ),
};
