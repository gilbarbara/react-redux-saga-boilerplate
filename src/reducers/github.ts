import { parseError } from 'modules/client';
import { createReducer } from 'modules/helpers';

import { ActionTypes, STATUS } from 'literals';

import { GitHubState, Topic } from 'types';

const topic: Topic = {
  cached: false,
  data: [],
  message: '',
  status: STATUS.IDLE,
  updatedAt: 0,
};

export const githubState: GitHubState = {
  topics: {},
  query: '',
};

export default {
  github: createReducer<GitHubState>(
    {
      [ActionTypes.GITHUB_GET_REPOS_REQUEST]: (draft, { payload }) => {
        const { query } = payload;
        draft.query = query;

        draft.topics[query] = draft.topics[query] || { ...topic };

        draft.topics[query].message = '';
        draft.topics[query].status = STATUS.RUNNING;
      },
      [ActionTypes.GITHUB_GET_REPOS_SUCCESS]: (draft, { meta, payload }) => {
        const { cached, query, updatedAt } = meta || {};

        draft.topics[query] = draft.topics[query] || { ...topic };

        draft.topics[query].cached = cached;
        draft.topics[query].data = payload;
        draft.topics[query].status = STATUS.SUCCESS;
        draft.topics[query].updatedAt = updatedAt;
      },
      [ActionTypes.GITHUB_GET_REPOS_FAILURE]: (draft, { meta, payload }) => {
        const { query } = meta || {};

        draft.topics[query] = draft.topics[query] || { ...topic };

        draft.topics[query].message = parseError(payload);
        draft.topics[query].status = STATUS.ERROR;
        draft.topics[query].updatedAt = 0;
      },
    },
    githubState,
  ),
};
