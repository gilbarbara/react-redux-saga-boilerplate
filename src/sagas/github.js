/**
 * @module Sagas/GitHub
 * @desc GitHub
 */

import { all, call, put, takeLatest } from 'redux-saga/effects';
import { request } from 'modules/client';

import { ActionTypes } from 'constants/index';

/**
 * Get Repos
 *
 * @param {Object} action
 *
 */
export function* getRepos({ payload }) {
  try {
    const response = yield call(
      request,
      `https://api.github.com/search/repositories?q=${payload.query}&sort=stars`,
    );
    yield put({
      type: ActionTypes.GITHUB_GET_REPOS_SUCCESS,
      payload: { data: response.items },
    });
  } catch (err) {
    /* istanbul ignore next */
    yield put({
      type: ActionTypes.GITHUB_GET_REPOS_FAILURE,
      payload: err,
    });
  }
}

/**
 * GitHub Sagas
 */
export default function* root() {
  yield all([takeLatest(ActionTypes.GITHUB_GET_REPOS, getRepos)]);
}
