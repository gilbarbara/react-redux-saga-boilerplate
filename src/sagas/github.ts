/**
 * @module Sagas/GitHub
 * @desc GitHub
 */

import { all, call, put, select, takeLatest } from 'redux-saga/effects';

import { request } from 'modules/client';
import { getUnixtime, hasValidCache } from 'modules/helpers';

import { ActionTypes } from 'literals';

import { StoreAction } from 'types';

/**
 * Get Repos
 *
 * @param {Object} action
 *
 */
export function* getRepos({ payload }: StoreAction) {
  const { query } = payload;

  try {
    const { data = [], updatedAt = 0 } = yield select(s => s.github.topics?.[query] || {});
    const hasCache = hasValidCache(updatedAt);
    let items;

    if (!hasCache) {
      ({ items = [] } = yield call(
        request,
        `https://api.github.com/search/repositories?q=${query}&sort=stars`,
      ));
    }

    yield put({
      type: ActionTypes.GITHUB_GET_REPOS_SUCCESS,
      payload: items || data,
      meta: { cached: hasCache, query, updatedAt: getUnixtime() },
    });
  } catch (err) {
    yield put({
      type: ActionTypes.GITHUB_GET_REPOS_FAILURE,
      payload: err,
      meta: { query },
    });
  }
}

/**
 * GitHub Sagas
 */
export default function* root() {
  yield all([takeLatest(ActionTypes.GITHUB_GET_REPOS_REQUEST, getRepos)]);
}
