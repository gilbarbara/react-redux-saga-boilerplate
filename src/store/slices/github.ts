import type { PayloadAction } from '@reduxjs/toolkit';
import { createSlice } from '@reduxjs/toolkit';

import { topic } from '~/config';

import { actionBody } from '~/modules/helpers';

import { STATUS } from '~/literals';

import { GitHubState } from '~/types';

interface GetReposSuccessMeta {
  cached: boolean;
  query: string;
  updatedAt: number;
}

export const githubState: GitHubState = {
  topics: {},
};

export const githubSlice = createSlice({
  name: 'github',
  initialState: githubState,
  reducers: {
    getRepos: (draft, { payload }: PayloadAction<string>) => {
      draft.topics[payload] = draft.topics[payload] || { ...topic };

      draft.topics[payload].message = '';
      draft.topics[payload].status = STATUS.RUNNING;
    },
    getReposSuccess: {
      reducer: (
        draft,
        { meta, payload }: PayloadAction<Record<string, any>[], string, GetReposSuccessMeta>,
      ) => {
        const { cached = false, query = '', updatedAt = 0 } = meta || {};

        draft.topics[query] = draft.topics[query] || { ...topic };

        draft.topics[query].cached = cached;
        draft.topics[query].data = payload;
        draft.topics[query].status = STATUS.SUCCESS;
        draft.topics[query].updatedAt = updatedAt;
      },
      prepare: (payload: Record<string, any>[], meta: GetReposSuccessMeta) =>
        actionBody(payload, meta),
    },
    getReposFailure: {
      reducer: (draft, { meta, payload }: PayloadAction<string, string, { query: string }>) => {
        const { query = '' } = meta || {};

        draft.topics[query] = draft.topics[query] || { ...topic };

        draft.topics[query].message = payload;
        draft.topics[query].status = STATUS.ERROR;
        draft.topics[query].updatedAt = 0;
      },
      prepare: (payload: string, query: string) => actionBody(payload, { query }),
    },
  },
});

export const { getRepos, getReposFailure, getReposSuccess } = githubSlice.actions;
export default githubSlice.reducer;
