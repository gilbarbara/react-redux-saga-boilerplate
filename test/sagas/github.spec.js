import { expectSaga } from 'redux-saga-test-plan';

import github, { getRepos } from 'sagas/github';
import { ActionTypes } from 'constants/index';

jest.mock('modules/client', () => ({
  request: () => ({ items: [] }),
}));

describe('github', () => {
  it('should have the expected watchers', done =>
    expectSaga(github)
      .run({ silenceTimeout: true })
      .then(saga => {
        expect(saga).toMatchSnapshot();
        done();
      }));

  it('should have the repos saga', () =>
    expectSaga(getRepos, { payload: { query: 'react' } })
      .put({
        type: ActionTypes.GITHUB_GET_REPOS_SUCCESS,
        payload: {
          data: [],
        },
      })
      .run());
});
