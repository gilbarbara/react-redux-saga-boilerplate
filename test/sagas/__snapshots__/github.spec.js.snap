// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`github should have the expected watchers 1`] = `
Object {
  "fork": Array [
    Object {
      "@@redux-saga/IO": true,
      "FORK": Object {
        "args": Array [
          "GITHUB_GET_USERS",
          "@@redux-saga-test-plan/json/function/getUsers",
        ],
        "context": null,
        "fn": "@@redux-saga-test-plan/json/function/takeLatest",
      },
    },
    Object {
      "@@redux-saga/IO": true,
      "FORK": Object {
        "args": Array [
          "GITHUB_GET_USER",
          "@@redux-saga-test-plan/json/function/getUser",
        ],
        "context": null,
        "fn": "@@redux-saga-test-plan/json/function/takeLatest",
      },
    },
  ],
  "take": Array [
    Object {
      "@@redux-saga/IO": true,
      "TAKE": Object {
        "pattern": "GITHUB_GET_USERS",
      },
    },
    Object {
      "@@redux-saga/IO": true,
      "TAKE": Object {
        "pattern": "GITHUB_GET_USER",
      },
    },
  ],
}
`;
